lib/api/CDataFrameAnalyzer.cc

```diff
@@ -250,10 +255,49 @@ void CDataFrameAnalyzer::captureFieldNames(const TStrVec& fieldNames) {
         return;
     }
     if (m_DataFrame != nullptr && m_CapturedFieldNames == false) {
+
         TStrVec columnNames{fieldNames.begin() + m_BeginDataFieldValues,
                             fieldNames.begin() + m_EndDataFieldValues};
-        m_DataFrame->columnNames(columnNames);
-        m_DataFrame->categoricalColumns(m_AnalysisSpecification->categoricalFieldNames());
+
+        if (m_DataFrame->hasColumnNames()) {
+
+            // We take the view that missing fields are not fatal, since they may
+            // not be available for the new set, but extra fields are likely to
+            // indicate user error.
+
+            TSizeVec positions(columnNames.size());
+            std::iota(positions.begin(), positions.end(), 0);
+            maths::COrderings::simultaneousSort(columnNames, positions);
+
+            TStrVec originalColumnNames{m_DataFrame->columnNames()};
+            std::sort(originalColumnNames.begin(), originalColumnNames.end());
+
+            m_ColumnMap = std::make_unique<TSizeVec>();
+            m_ColumnMap->reserve(columnNames.size());
+
+            TStrVec extraColumnNames;
+            std::set_difference(columnNames.begin(), columnNames.end(),
+                                originalColumnNames.begin(), originalColumnNames.end(),
+                                std::back_inserter(extraColumnNames));
+            if (extraColumnNames.empty() == false) {
+                HANDLE_FATAL(<< "Input error: supplying additional columns "
+                             << core::CContainerPrinter::print(extraColumnNames) << ".");
+            }
+
+            for (const auto& name : m_DataFrame->columnNames()) {
+                auto i = std::lower_bound(columnNames.begin(), columnNames.end(), name);
+                if (i == columnNames.end() || *i != name) {
+                    LOG_WARN(<< "Missing column '" << name << "'");
+                    m_ColumnMap->push_back(columnNames.size());
```

Comment:
This seems to denote missing values with `columnNames.size()`. However, in a comment in `CDataFrameAnalyzer.h` we have ```// This has values: -2 (unset), -1 (missing), >= 0 (control field index).```. Should we not be denoting missing values with `-1`? Or this refers to something else?

Reply:
I agree it would be more consistent in `CDataFrameAnalyzer`, but I also think it is more natural where it's used in `CDataFrame::parseAndWriteRow` to make this a vector of unsigned ints and denote missing as I have done. I guess this is a tradeoff and I don't have strong opinions, if you find the suggested version clearer I'm happy to change: see [e7be3e6](https://github.com/elastic/ml-cpp/pull/2052/commits/e7be3e6d2ddfdd751392090d3f6bed2585b11fce).

Reply:
I find it easier to read and understand without e7be3e6. Let it be inconsistent :-)